### PR TITLE
Code Updated dated 17 feb

### PR DESCRIPTION
**Pull request by abhishekkrsp**
# Updated
- Timer
- MVC
- StackView
- Added one enum status case matched
- Resolved Alternate Card Issue
- Optimised updateViewModel()
- Added assets using extension
- made one separate file to handle emojis
- emoji packet added
- Resolved duplicate execution

### CODE REVIEW

File: card-picker/card-picker/ViewController.swift [hunk 1]
```diff
@@ -7,88 +7,76 @@
...
+        }
         flipCount += 1
-        if let index = cardButtons.firstIndex(of: sender) {
-            print("button \(index) is clicked")
-            flipCard(with: emojis[index], on: sender)
+        if let cardNumber = cardButtons.firstIndex(of: sender) {
+            game.chooseCard(at: cardNumber)
+            updateViewFromModel()
+            game.matchCard(at: cardNumber)
+            Timer.scheduledTimer(timeInterval: 0.25, target: self, selector: #selector(updateViewFromModel), userInfo: nil, repeats: false)
```
Comment (priyal-p) [inline card-picker/card-picker/ViewController.swift]: In line 51 we are calling updateViewFromModel(), and this method will be executed after Timer is over, Do we need both?

File: card-picker/card-picker/ViewController.swift [hunk 2]
```diff
@@ -7,88 +7,76 @@
...
-            }
+    }
+    
+    var emojiChoices = ["🥳", "😜", "🤝", "👽", "😋", "✌️",]
+
+
+    var emoji = [Int: String]()
+    func emoji(for card: Card) -> String {
+        if emoji[card.identifier] == nil, emojiChoices.count > 0 {
+            let randomIndex = Int(arc4random_uniform(UInt32(emojiChoices.count)))
```
Comment (priyal-p) [inline card-picker/card-picker/ViewController.swift]: Could you please try using Int.random() function for this usecase.

File: card-picker/card-picker/ViewController.swift [hunk 3]
```diff
@@ -7,88 +7,76 @@
...
     
-   
     @IBOutlet var cardButtons: [UIButton]!
-    
-    
+    func switchCardState(with button: UIButton, for card: Card) {
+        switch card.state {
+        case .faceUp:
+            button.setTitle(emoji(for: card), for: .normal);
+            button.backgroundColor = #colorLiteral(red: 1, green: 1, blue: 1, alpha: 1)
```
Comment (priyal-p) [inline card-picker/card-picker/ViewController.swift]: Add the color in Assets, same for below lines.

File: card-picker/card-picker/Card.swift [hunk 4]
```diff
@@ -7,8 +7,23 @@
...
+
 struct Card {
-    var isFaceUp = false
-    var isRemoved = false
-    var emoji: String
+    var state: CardStates = .faceDown
+    var emoji = String()
+    var identifier: Int
+    static var identifierFactory = 0
+    static var lock = true
```
Comment (priyal-p) [inline card-picker/card-picker/Card.swift]: If we are not using lock, Please consider removing it.

File: card-picker/card-picker/ViewController.swift (hunk 1 above)
Comment (priyal-p) [reply card-picker/card-picker/ViewController.swift]: We can remove `userInfo` and `repeats` as their default value is same.

File: card-picker/card-picker/ViewController.swift [hunk 5]
```diff
@@ -7,88 +7,76 @@
...
-    func flipCard(with emoji: String, on button: UIButton) {
-        print("emoji: \(button.currentTitle)")
-        if button.currentTitle != nil {
-            faceUp -= 1
-            button.setTitle(nil, for: .normal)
-            button.backgroundColor = #colorLiteral(red: 0.5791940689, green: 0.1280144453, blue: 0.5726861358, alpha: 1)
-
-            
+    @objc func updateViewFromModel() {
+        for index in cardButtons.indices {
```
Comment (priyal-p) [inline card-picker/card-picker/ViewController.swift]: Do we need to iterate over all cardButtons indices here?

**Review by priyal-p — COMMENTED** [review]
Also, we can avoid to commit .DS_Store, and .xcuserstate file. You can consider adding these in `git ignore` file.

File: card-picker/card-picker/ViewController.swift (hunk 1 above)
Comment (abhishekkrsp) [reply card-picker/card-picker/ViewController.swift]: > In line 51 we are calling updateViewFromModel(), and this method will be executed after Timer is over, Do we need both?

Yes one for instant click change (when 3rd card is picked) after click and other with timer updation.

Comment (abhishekkrsp) [reply card-picker/card-picker/ViewController.swift]: > We can remove `userInfo` and `repeats` as their default value is same.

Sure we can change.

File: card-picker/card-picker/Card.swift (hunk 4 above)
Comment (abhishekkrsp) [reply card-picker/card-picker/Card.swift]: Ok

File: card-picker/card-picker/ViewController.swift (hunk 5 above)
Comment (abhishekkrsp) [reply card-picker/card-picker/ViewController.swift]: > No, we can optimise.

File: card-picker/card-picker/ViewController.swift [hunk 6]
```diff
@@ -7,88 +7,84 @@
...
+    var flipIndex: (Int?, Int?)
     
-    
-    @IBAction func touchCard(_ sender: UIButton) {
-        flipCount += 1
-        if let index = cardButtons.firstIndex(of: sender) {
-            print("button \(index) is clicked")
-            flipCard(with: emojis[index], on: sender)
+    func switchCardState(with index: Int?) {
+        if index == nil {
```
Comment (priyal-p) [inline card-picker/card-picker/ViewController.swift]: We can use guard let here

Comment (abhishekkrsp) [reply card-picker/card-picker/ViewController.swift]: > We can use guard let here

Ok

**Comment by abhishekkrsp** [comment]
Updated CodeBase

**Pull request merged**